### DISCUSSION TITLE
Timer and a while loop - does the timer interrupt execution

### DISCUSSION THREAD

**Discussion by orangepineapple** [ANLAB-KAIST/KENSv3#153] [Project Q&A 2. Transport Layer (TCP)]
I read in an older post that the timer will not work if you call the timer and then enter a conditional while loop that could not end. For example the below code:

 #78 

```
addTimer(1 second)
while( ack_not_received == True){
  do_nothing() 
} 

timeCallback(){
  ack_not_recevied = False;
  timed_out = True;
}
```

I'm wondering if there is a specific reason that the timer callback cannot interrupt the while loop. Also if this style of timeout doesn't work, I'm not sure how we would implement a packet time out. Is there any hint you could give in that case?

**Comment by Pusnow** [comment]
KENS follows single-threaded event-driven model, so you should avoid busy-waiting (`while( ack_not_received == True)`) to proceed.

Rather than busy-waiting, the original function must return, and later the remaining tasks should be handled in `timerCallback()`.

**Reply by orangepineapple** [reply]
Ok, I see. So it would be ok to design out function to return after a constant amount of time on each call. Regardless of when the response arrives. I also noticed there was not specification for how long the timer should be, is there a recommended timeout?

**Reply by Pusnow** [reply]
It depends on types of timeout.

For retransmission timeout, see:

https://github.com/ANLAB-KAIST/KENSv3/blob/eba97b6c81103e365e14e3c13194ccbc2a5966f3/Documentation/rfc/rfc793.txt#L2609-L2634

**Reply by orangepineapple** [reply]
gotcha, and I'm assuming we use c++ chrono to measure the RTT of the acks
